File: knowledge-packages/m13-reference-providers/providers/node/pinecone-runtime.mjs
import {
  assertAttestedRequest,
  capabilities,
  normalizeExternalRows,
  pineconeQueryBody,
  pineconeMatchesToRows,
  providerConfig,
  withTypedFailure,
} from "./lib.mjs";

const { serveKnowledgeRuntimeProcess } = await import("@agentpm/sdk");
if (typeof serveKnowledgeRuntimeProcess !== "function") {
  throw new Error(
    "Installed @agentpm/sdk does not export serveKnowledgeRuntimeProcess. Install a newer SDK release, or build and link the local agentpm-sdk-node checkout for pre-publish milestone testing.",
  );
}

const config = providerConfig("pinecone-reference");

function requireEnv(name) {
  const value = process.env[name];
  if (!value) throw new Error(`${name} is required`);
  return value;
}

async function embedQuery(text) {
  const response = await fetch("https://api.openai.com/v1/embeddings", {
    method: "POST",
    headers: {
      Authorization: `Bearer ${requireEnv("OPENAI_API_KEY")}`,
      "Content-Type": "application/json",
    },
    body: JSON.stringify({ model: config.embeddingModel, input: text }),
  });
  if (!response.ok) {
    throw new Error(
      `OpenAI embedding request failed with HTTP ${response.status}`,
    );
  }
  const payload = await response.json();
  const vector = payload.data?.[0]?.embedding;
  if (!Array.isArray(vector) || vector.length !== config.dimensions) {
    throw new Error(
      `OpenAI embedding dimensions did not match ${config.dimensions}`,
    );
  }
  return vector;
}

async function retrieve(request) {
  return withTypedFailure(request, "pinecone_query_failed", async () => {
    assertAttestedRequest(request, config);
    const vector = await embedQuery(request.query);
    const response = await fetch(
      `${requireEnv("PINECONE_INDEX_HOST").replace(/\/$/, "")}/query`,
      {
        method: "POST",
        headers: {
          "Api-Key": requireEnv("PINECONE_API_KEY"),
          "Content-Type": "application/json",
        },
        body: JSON.stringify(
          pineconeQueryBody(
            request,
            vector,
            config,
            process.env.PINECONE_NAMESPACE || "",
          ),
        ),
      },
    );
    if (!response.ok) {
      throw new Error(`Pinecone query failed with HTTP ${response.status}`);
    }
    const payload = await response.json();
    return normalizeExternalRows(
      request,
      pineconeMatchesToRows(
        Array.isArray(payload.matches) ? payload.matches : [],
        config,
      ),
      { returnCitationsDefault: config.returnCitationsDefault },
    );
  });
}

await serveKnowledgeRuntimeProcess(
  config.runtimeId,
  retrieve,
  capabilities(config),
);

File: knowledge-packages/m13-reference-providers/providers/python/pinecone_runtime.py
from __future__ import annotations

import json
import os
import urllib.error
import urllib.request
from typing import TYPE_CHECKING, Any

if TYPE_CHECKING:
    from agentpm import KnowledgeRuntimeRequest, KnowledgeRuntimeResult
else:
    KnowledgeRuntimeRequest = dict[str, Any]
    KnowledgeRuntimeResult = dict[str, Any]

from reference_runtime import (
    assert_attested_request,
    capabilities,
    failure_result,
    normalize_external_rows,
    openai_embed,
    pinecone_query_body,
    pinecone_matches_to_rows,
    provider_config,
)

CONFIG = provider_config("pinecone-reference")


def require_env(name: str) -> str:
    value = os.environ.get(name)
    if not value:
        raise RuntimeError(f"{name} is required")
    return value


def retrieve(request: KnowledgeRuntimeRequest) -> KnowledgeRuntimeResult:
    try:
        assert_attested_request(request, CONFIG)
        vector = openai_embed(request["query"], CONFIG)
        body = json.dumps(
            pinecone_query_body(
                request,
                vector,
                CONFIG,
                os.environ.get("PINECONE_NAMESPACE", ""),
            )
        ).encode("utf-8")
        url = require_env("PINECONE_INDEX_HOST").rstrip("/") + "/query"
        pinecone_request = urllib.request.Request(
            url,
            data=body,
            headers={
                "Api-Key": require_env("PINECONE_API_KEY"),
                "Content-Type": "application/json",
            },
            method="POST",
        )
        try:
            with urllib.request.urlopen(pinecone_request) as response:
                payload = json.loads(response.read().decode("utf-8"))
        except urllib.error.HTTPError as exc:
            raise RuntimeError(f"Pinecone query failed with HTTP {exc.code}") from exc
        return normalize_external_rows(
            request,
            pinecone_matches_to_rows(payload.get("matches", []), CONFIG),
            return_citations_default=CONFIG.return_citations_default,
        )
    except Exception as exc:
        return failure_result(request, "pinecone_query_failed", str(exc))


def main() -> None:
    try:
        from agentpm import serve_knowledge_runtime_process
    except ImportError as exc:
        raise RuntimeError(
            "Installed agentpm SDK does not export serve_knowledge_runtime_process. "
            "Install a newer SDK release, or run this provider with "
            "`uv run --with-editable ../../../agentpm-sdk-python python "
            "providers/python/pinecone_runtime.py` for pre-publish milestone testing."
        ) from exc

    serve_knowledge_runtime_process(CONFIG.runtime_id, retrieve, capabilities(CONFIG))


if __name__ == "__main__":
    main()
